### PR TITLE
#903 Use timeline events and constraints only if exist

### LINKED ISSUE #903: Build and write in the different output files only when necessary
As of today, the jobs file enables to export or not a timeline, export or not the constraints, export or not the logs and so on. Except for logs (which have been recently refactorized and improved), the other infos are built even if they won't be exported later on.

The development targets at avoiding to construct and fill the objects if they won't be exported later on.

### PR DESCRIPTION
**Pull request by lecourtoisflo**
closes #903

### CODE REVIEW

**Comment by lecourtoisflo** [comment]
at most 40s gain on big tests

File: dynawo/sources/Simulation/DYNSimulation.cpp [hunk 1]
```diff
@@ -281,6 +281,9 @@ void
 Simulation::configureConstraintsOutputs() {
   // Constraints settings
   if (jobEntry_->getOutputsEntry()->getConstraintsEntry()) {
+    stringstream pid_string;
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulation.cpp:284]: Maybe we can remove the timeline from solvers since it is not used?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Simulation/DYNSimulation.cpp (hunk 1 above)
Comment (lecourtoisflo) [reply dynawo/sources/Simulation/DYNSimulation.cpp:284]: it is used at least by solverSIM in solveWithStepRecalculation

Comment (rosiereflo) [reply dynawo/sources/Simulation/DYNSimulation.cpp:284]: OK then there should be an update on the solver side right?

**Review by rosiereflo — APPROVED** [review]

File: dynawo/sources/Modeler/Common/DYNSubModel.cpp [hunk 2]
```diff
@@ -127,6 +127,11 @@ SubModel::setTimeline(const shared_ptr<Timeline>& timeline) {
   timeline_ = timeline;
 }
 
+boost::shared_ptr<timeline::Timeline>
+SubModel::getTimeline() const {
```
Comment (gautierbureau) [inline dynawo/sources/Modeler/Common/DYNSubModel.cpp:131]: I would prefer bool hasTimeline(); if this getter is only used to test if pointer is NULL, @rosiereflo what is your take on this ?

**Review by gautierbureau — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/Common/DYNSubModel.cpp (hunk 2 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp:131]: I agree

Comment (e2l-su) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp:131]: I agree too, but getTimeline() is used in new version if DYNAddEvent(), the macro, to gain access to the timeline :
DYNMacrosMessage.h:            timeline::Timeline& tl(*(pNetwork)->getTimeline());

Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp:131]: Hi, Florent has taken over from your original branch and simplified the macros, so I think that they do not use this method anymore. If I'm not mistaken with the current version Gautier is right we can replace this method with a hasTimeline

Comment (lecourtoisflo) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp:131]: correct

File: dynawo/sources/Modeler/Common/DYNSubModel.h [hunk 3]
```diff
@@ -750,11 +750,11 @@ class SubModel {
       const std::string& modelType = "");
 
   /**
-   * @brief Retrieves the constraint collection
+   * @brief Determines i the model has a constraint collection
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNSubModel.h:753]: i => if?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/Common/DYNSubModel.h (hunk 3 above)
Comment (lecourtoisflo) [reply dynawo/sources/Modeler/Common/DYNSubModel.h]: OK

**Review by rosiereflo — APPROVED** [review]

**Review by gautierbureau — APPROVED** [review]

File: dynawo/sources/Modeler/Common/DYNSubModel.cpp (hunk 2 above)
Comment (e2l-su) [reply dynawo/sources/Modeler/Common/DYNSubModel.cpp]: Thank you Florent :) I don't follow anymore. Did you see I added an unitary test for the macro? See you

**Pull request merged**